### DISCUSSION TITLE
LDAP Integration issue

### DISCUSSION THREAD

**Discussion by GFM-Network** [Ylianst/MeshCentral#6262] [Issues]
I'm trying to setup the LDAP on our meshcentral at work but everytime I change the ldap auth settings it starts with errors and stop working the web ui, can someone help me spot what I'm missing on my config.json file?
![image](https://github.com/user-attachments/assets/9977c1c7-b1ef-436a-bd7f-76be63b8c644)

```
{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "settings": {
    "MongoDb": "mongodb://127.0.0.1:27017/meshcentral",
    "cert": "hidden",
    "_WANonly": true,
    "_LANonly": true,
    "_sessionKey": "MyReallySecretPassword1",
    "_port": 443,
    "_aliasPort": 443,
    "_redirPort": 80,
    "_redirAliasPort": 80
  },
  "domains": {
    "": {
      "auth": "ldap",
      "ldapUserGroups": "memberOf",
      "ldapSyncWithUserGroups": {
      "filter": [ "hidden" ]
      },
      "ldapoptions": {
        "url": "ldap://10.58.254.71:389/",
        "bindDN": "hidden",
        "bindCredentials": "hidden",
        "searchBase": "hidden",
        "searchFilter": "(name={{username}})",
        "reconnect": true
      },
      "ldapUserBinaryKey": "objectSid",
      "ldapUserName": "sAMAccountName",
      "ldapUserEmail": "mail",
      "ldapUserPhoneNumber": "telephoneNumber",
      "ldapUserImage": "thumbnailPhoto",
      "ldapSaveUserToFile": "C:\\tmp\\ldapuser-save.txt"
  }
 },
  "_letsencrypt": {
    "_comment_": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before trying Let's Encrypt.",
    "email": "myemail@mydomain.com",
    "names": "myserver.mydomain.com",
    "skipChallengeVerification": true,
    "production": false
  }
}
```

I've made some changes now the webui open but I get this error:
https://imgur.com/LbqaFwp

**Comment by si458** [comment]
One thing to try is run meshcentral in debug mode and see what it does when it logs in and out
`node node_modules/meshcentral --debug`
Or
If docker add environment variable 
`ARGS: --debug`

**Reply by GFM-Network** [reply]
Where do I run that command? 

![image](https://github.com/user-attachments/assets/6a44df73-5ffe-4ac8-bc24-f5023dbaaecd)

**Reply by si458** [reply]
sorry been out at work, how did you install meshcentral? 
you would generally run the `node node_modules/meshcentral --debug` command from base folder
which in ur case would be the `/var/opt/meshcentral` folder
just make sure meshcentral isnt running first `systemctl status meshcentral` -> `systemctl stop meshcentral`

**Reply by GFM-Network** [reply]
Here's the logs when I try to login with a ldap user:


HTTPHEADERS: POST /login {
  host: 'redacted',
  'x-forwarded-scheme': 'https',
  'x-forwarded-proto': 'https',
  'x-forwarded-for': '10.58.72.158',
  'x-real-ip': '10.58.72.158',
  'content-length': '61',
  'cache-control': 'max-age=0',
  'sec-ch-ua': '"Not/A)Brand";v="8", "Chromium";v="126", "Brave";v="126"',
  'sec-ch-ua-mobile': '?0',
  'sec-ch-ua-platform': '"Windows"',
  'upgrade-insecure-requests': '1',
  origin: 'null',
  'content-type': 'application/x-www-form-urlencoded',
  'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36',
  accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
  'sec-gpc': '1',
  'accept-language': 'en-US,en;q=0.6',
  'sec-fetch-site': 'same-origin',
  'sec-fetch-mode': 'navigate',
  'sec-fetch-user': '?1',
  'sec-fetch-dest': 'document',
  'accept-encoding': 'gzip, deflate, br, zstd',
  priority: 'u=0, i',
  cookie: 'xid=e30=; xid.sig=zuDN3EHulnrChGktskSzgpHe7Sh2QU5nhdA2zuwuB8ywEpXxhriyL7bnrJtSyqbK'
}
WEBREQUEST: (10.58.64.55) /login
WEB: handleRootPostRequest, action: login
LDAP: LDAP Error: no such user: "redacted"
AUTHLOG: Failed password for redacted from 10.58.64.55 port 51770, Browser: Chrome/126.0.0.0, OS: Windows/10

**Reply by GFM-Network** [reply]
on search base do I need to specify the DN/OU where the users are located or just the base domain tree DN works?

**Reply by si458** [reply]
ok so the is something wrong with your LDAP because its say `LDAP: LDAP Error: no such user: "redacted"`
so it cant find the user
whats ur LDAP provider?

**Reply by si458** [reply]
this is my example ldap i use and im using openldap
```
 "": {
      "auth": "ldap",
      "LDAPOptions": {
        "url": "ldap://localhost:389",
        "bindDN": "cn=admin,dc=example,dc=org",
        "bindCredentials": "adminpassword",
        "searchBase": "ou=StaffUsers,dc=example,dc=org",
        "searchFilter": "(uid={{username}})",
        "reconnect": true,
        "tlsOptions": {
          "rejectUnauthorized": false
        }
      },
      "ldapUserKey": "uid",
      "ldapUserName": "uid",
      "ldapUserRealName": "displayName",
      "_ldapSiteAdminGroups": "500",
      "_ldapUserGroups": "gidNumber",
      "ldapSyncWithUserGroups": true,
      "ldapSaveUserToFile": "../meshcentral-data/ldap.txt"
    },
```

**Reply by GFM-Network** [reply]
Changed to your settings this is the result :(

![image](https://github.com/user-attachments/assets/06ec7e0e-b3ab-4491-b9d7-5a805db1cc56)

**Reply by si458** [reply]
What's ur ldap provider? 
Windows active directory? Jumpcloud? Openldap?
U need to config the ldap to look up correctly the user's then user groups etc

**Reply by GFM-Network** [reply]
I even used a software to make sure I was getting the correct path of OU and CN.

I'm using active directory:

path to the service user to use in meshcentral: 
"CN=Meshcentral Service,OU=Redacted Services,OU=Redacted2,OU=Redacted1,DC=Redacted,DC=org,DC=com"

searchbase:
"DC=Redacted,DC=org,DC=com"

I even tried using the same user and path on a nextcloud instance that I have and it works fine, meshcentral is driving me nuts xD

**Reply by si458** [reply]
I will have a play 2moz with my works AD see if I can figure out what the values need to be for u.
Are u logging in with the username or there email address?

**Reply by GFM-Network** [reply]
I tried login with the user logon that's what I use for other services:

![image](https://github.com/user-attachments/assets/e1d686db-eb17-40b7-bbd9-fac02354b108)

**Reply by thermionic** [reply]
As you have used 
`"ldapUserName": "sAMAccountName",`
Then you need to use the "User logon name (pre-Windows 2000)"

**Reply by GFM-Network** [reply]
but both: user logon name / user logon name (pre-windows 200): 
have the same values, I shouldn't use "sAMAccountName" then?

**Reply by thermionic** [reply]
the other logon name would be the UPN which also requires the domain such as user@domain.tld

presuming that you are using LDAP on 389 and not LDAPS, I would run a packet capture on either the MC server or the AD DC and see if you see what is being sent "on the wire".
